src/main/java/org/swisspush/reststorage/FileSystemDirLister.java

```diff
@@ -0,0 +1,137 @@
+            log.trace("worker-thread says bye.");
+        }, event -> {
+            log.trace("Welcome back on eventloop-thread.");
+            if (log.isDebugEnabled()) {
+                final long durationMillis = System.currentTimeMillis() - startTimeMillis;
+                log.debug("List directory contents of '{}' took {}ms", path, durationMillis);
+            }
+            if (event.failed()) {
+                log.error("Directory listing failed.", event.cause());
+                final Resource erroneousResource = new Resource() {{
+                    // Set fields according to documentation in Resource class.
+                    name = Paths.get(path).getFileName().toString();
+                    exists = false;
+                    error = rejected = invalid = true;
+                    errorMessage = invalidMessage = event.cause().getMessage();
+                }};
+                handler.handle(erroneousResource);
+            } else {
+                handler.handle((Resource) event.result());
+            }
+        });
+        log.trace("Work delegated.");
+    }
+
+    private void listDirBlocking(String path, int offset, int count, Future<CollectionResource> future) {
+        //
+        // HINT: This method gets executed on a worker thread!
+        //
+        // Convert String to Path
+        final Path searchPath = Paths.get(canonicalizeVirtualPath(path));
+        // Prepare our result.
+        final CollectionResource collection = new CollectionResource() {{
+            items = new ArrayList<>(128);
+        }};
+        final String fullPath = canonicalizeVirtualPath(path);
+        try (Stream<Path> source = Files.list(searchPath)) {
+            source.forEach(entry -> {
+                final String entryName = entry.getFileName().toString();
+                log.trace("Processing entry '{}'", entryName);
+                if (".tmp".equals(entryName) && fullPath.length() == root.length()) {
+                    // Ignore hidden '/.tmp/' directory.
+                    return;
+                }
+                // Create resource representing currently processed directory entry.
+                final Resource resource;
+                if (Files.isDirectory(entry)) {
+                    resource = new CollectionResource();
+                } else if (Files.isRegularFile(entry)) {
+                    resource = new DocumentResource();
+                } else {
+                    resource = new Resource();
+                    resource.exists = false;
+                }
+                resource.name = entryName;
+                collection.items.add(resource);
+            });
+        } catch (IOException e) {
+            future.fail(e);
+            return;
+        }
+        Collections.sort(collection.items);
+        // Don't know exactly what we do here now. Seems we check 'limit' for a range request.
```

Comment:
Do not keep those comments in code. Clear things out before submitting the PR.

Reply:
_Comments are documentation_: It tooks some time until I got what this code actually is for. Without the comment, every dev needs to get the point of this code over and over again with the risk to miss-interpret it.

I suggest to reword the comment instead removing it. If we do so: Are my assumptions in comment correct?